crates/worker/src/partition_processor_manager/snapshot_task.rs

```diff
@@ -0,0 +1,149 @@
+// Copyright (c) 2024 - Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::path::PathBuf;
+use std::time::SystemTime;
+
+use tokio::sync::{oneshot, watch};
+use tracing::{debug, warn};
+
+use restate_core::worker_api::{SnapshotError, SnapshotResult};
+use restate_partition_store::snapshots::{
+    LocalPartitionSnapshot, PartitionSnapshotMetadata, SnapshotFormatVersion,
+};
+use restate_partition_store::PartitionStoreManager;
+use restate_types::identifiers::{PartitionId, SnapshotId};
+use restate_types::logs::Lsn;
+
+/// Handle to an outstanding [`SnapshotPartitionTask`] that has been spawned, including a reference
+/// to notify the requester.
+pub struct PendingSnapshotTask {
+    pub sender: oneshot::Sender<SnapshotResult>,
+}
+
+/// Creates a partition store snapshot along with Restate snapshot metadata.
+pub struct SnapshotPartitionTask {
+    pub cluster_name: String,
+    pub node_name: String,
+    pub partition_id: PartitionId,
+    pub snapshot_base_path: PathBuf,
+    pub partition_store_manager: PartitionStoreManager,
+    pub archived_lsn_sender: watch::Sender<Option<Lsn>>,
+}
+
+impl SnapshotPartitionTask {
+    pub async fn create_snapshot(self) -> Result<PartitionSnapshotMetadata, SnapshotError> {
+        debug!(
+            partition_id = %self.partition_id,
+            "Creating partition snapshot"
+        );
+
+        let result = create_snapshot_inner(
+            self.partition_store_manager,
+            self.cluster_name,
+            self.node_name,
+            self.partition_id,
+            self.snapshot_base_path,
+            self.archived_lsn_sender,
+        )
+        .await;
+
+        match result {
+            Ok(metadata) => {
+                debug!(
+                    partition_id = %self.partition_id,
+                    snapshot_id = %metadata.snapshot_id,
```

Comment:
If the `SnapshotPartitionTask` created the `SnapshotId` as part of the instantiation, then you could also include the `snapshot_id` in the `create_snapshot` span.